cirq-google/cirq_google/named_topologies.py

```diff
@@ -0,0 +1,259 @@
+
+    nx.draw_networkx(graph, pos=pos, ax=ax, **kwargs)
+    ax.axis('equal')
+    return pos
+
+
+@dataclass(frozen=True)
+class LineTopology(NamedTopology):
+    """A 1D linear topology.
+
+    Node indices are contiguous integers starting from 0 with edges between
+    adjacent integers.
+
+    Args:
+        n_nodes: The number of nodes in a line.
+    """
+
+    n_nodes: int
+
+    # https://github.com/python/mypy/issues/1362
+    @property  # type: ignore
+    @cache
+    def name(self) -> str:
+        """The name of this topology: {n_nodes}-line"""
+        return f'{self.n_nodes}-line'
+
+    @property  # type: ignore
+    @cache
+    def graph(self) -> nx.Graph:
+        """The graph of this topology.
+
+        Node indices are contiguous integers starting from 0 with edges between
+        adjacent integers.
+        """
+        return nx.from_edgelist(
+            [(i1, i2) for i1, i2 in zip(range(self.n_nodes), range(1, self.n_nodes))]
+        )
+
+    def draw(self, ax=None, cartesian: bool = True, **kwargs) -> Dict[Any, Tuple[int, int]]:
+        """Draw this graph.
+
+        Args:
+            ax: Optional matplotlib axis to use for drawing.
+            cartesian: If True, draw as a horizontal line. Otherwise, draw on a diagonal.
+            kwargs: Additional arguments to pass to `nx.draw_networkx`.
+        """
+        g2 = nx.relabel_nodes(self.graph, {n: (n, 1) for n in self.graph.nodes})
+        return draw_gridlike(g2, ax=ax, cartesian=cartesian, **kwargs)
+
+    def _json_dict_(self):
+        return dataclass_json_dict(self, namespace='cirq.google')
+
+
+@dataclass(frozen=True)
+class DiagonalRectangleTopology(NamedTopology):
+    width: int
+    height: int
+
+    def __post_init__(self):
+        if self.width <= 0:
+            raise ValueError("Width must be a positive integer")
+        if self.height <= 0:
+            raise ValueError("Height must be a positive integer")
+
+    @property  # type: ignore
+    @cache
+    def name(self) -> str:
+        return f'{self.width}-{self.height}-diagonal-rectangle'
+
+    @property  # type: ignore
+    @cache
+    def graph(self) -> nx.Graph:
+        g = nx.Graph()
+        # construct a "diagonal rectangle graph" whose width and height
+        # set the number of rows of 'central' qubits, each of which has
+        # four neighbors in each cardinal direction
+        for megarow in range(self.height):
+            for megacol in range(self.width):
+                y = megacol + megarow
+                x = megacol - megarow
+                g.add_edge((x, y), (x - 1, y))
+                g.add_edge((x, y), (x, y - 1))
+                g.add_edge((x, y), (x + 1, y))
+                g.add_edge((x, y), (x, y + 1))
+        return g
+
+    @property  # type: ignore
+    @cache
+    def n_nodes(self) -> int:
+        return 2 * self.width * self.height + self.width + self.height + 1
```

Comment:
Wait, what?  Why isn't this width*height?  This is not what I was expecting.

Reply:
hopefully the class docstring should clarify what topology we're actually creating here. Our "unit cell" is four qubits in a square connected to one central qubit, so `n_nodes = ((1/4)*4 + 1)*w*h + boundary_effects`